internal/core/model/adapters.rs

```diff
@@ -413,3 +415,510 @@ fn test_filter_model() {
     assert_eq!(filter.row_data(4).unwrap(), 8);
     assert_eq!(filter.row_count(), 5);
 }
+
+trait SortHelper<D> {
+    fn sort(&self, lhs: &D, rhs: &D) -> std::cmp::Ordering;
+}
+
+struct AscendingSortHelper;
+
+impl<D> SortHelper<D> for AscendingSortHelper
+where
+    D: std::cmp::Ord,
+{
+    fn sort(&self, lhs: &D, rhs: &D) -> std::cmp::Ordering {
+        lhs.cmp(rhs)
+    }
+}
+
+struct FnSortHelper<F, D>
+where
+    F: FnMut(&D, &D) -> std::cmp::Ordering + 'static,
+{
+    sort_function: RefCell<F>,
+    _type: PhantomData<D>,
+}
+
+impl<F, D> SortHelper<D> for FnSortHelper<F, D>
+where
+    F: FnMut(&D, &D) -> std::cmp::Ordering + 'static,
+{
+    fn sort(&self, lhs: &D, rhs: &D) -> std::cmp::Ordering {
+        (self.sort_function.borrow_mut())(lhs, rhs)
+    }
+}
+
+struct SortModelInner<M>
+where
+    M: Model + 'static,
+{
+    wrapped_model: M,
+    sort_helper: Box<dyn SortHelper<M::Data>>,
+    // This vector saves the indices of the elements that are not filtered out
+    mapping: RefCell<Vec<usize>>,
+    notify: ModelNotify,
+}
+
+impl<M> SortModelInner<M>
+where
+    M: Model + 'static,
+{
+    fn build_mapping_vec(&self) {
+        self.mapping.borrow_mut().clear();
+        self.mapping.borrow_mut().extend((0..self.wrapped_model.row_count()).into_iter());
+        self.mapping.borrow_mut().sort_by(|lhs, rhs| {
+            self.sort_helper.sort(
+                &self.wrapped_model.row_data(*lhs).unwrap(),
+                &self.wrapped_model.row_data(*rhs).unwrap(),
+            )
+        });
+    }
+}
+
+impl<M> ModelChangeListener for SortModelInner<M>
+where
+    M: Model + 'static,
+{
+    fn row_changed(&self, row: usize) {
+        let removed_index = self.mapping.borrow().binary_search(&row).unwrap();
+        self.mapping.borrow_mut().remove(removed_index);
+
+        let insertion_index = if let Some(insertion_index) = self
```

Comment:
I think instead of a linear search, you can also do a binary search for the row that's suitable for insertion. Something like this (untested and badly formatted):

```rust
let changed_data = self.wrapped_model.borrow().row_data(row).unwrap();
let insertion_index = self.mapping.borrow().partition_point(|existing_row|
    self.sort_helper.sort(&self.wrapped_model.borrow().row_data(existing_row).unwrap(),
                                       changed_data) == std::cmp::Ordering::Less
);
```

Reply:
It works and it's done.